### PR TITLE
ITE: drivers/peci: Add PECI driver module of ITE IT8xxx2

### PR DESCRIPTION
**Pull request by bjchentc**
Added the PECI driver module functionally tested with the samples/drivers/peci.

Signed-off-by: BJ Chen <bj.chen@ite.com.tw>

### CODE REVIEW

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts [hunk 1]
```diff
@@ -15,6 +15,7 @@
 
 	aliases {
 		i2c-0 = &i2c0;
+		peci-0 = &peci0;
```
Comment (mbolivar-nordic) [inline boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:18]: Why do you need this? These sorts of aliases are generally copy/pasted from older board devicetrees that were written before we had APIs like `DT_NODELABEL`. I think it should probably be removed from this PR. I'm not sure why there's an `i2c-0` alias above, either, but that is not related to this PR.

File: drivers/peci/peci_ite_it8xxx2.c [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+	IRQ_CONNECT(DT_INST_IRQN(0),
+	0,
+	peci_it8xxx2_isr,
+	DEVICE_DT_INST_GET(0),
+	0);
+
+	return 0;
+}
+
+DEVICE_DT_INST_DEFINE(0,
```
Comment (mbolivar-nordic) [inline drivers/peci/peci_ite_it8xxx2.c:364]: I'm guessing this is a single-instance driver because there is only one such peripheral on currently supported SoCs?

If so, can you please add a `BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1, ...)` line here too, and a comment explaining what's going on?

If not, please make this a multi-instance driver; we are well past the time in the project where we should be adding single-instance drivers when there are potentially multiple hardware instances.

File: drivers/peci/peci_ite_it8xxx2.c [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+	case PECI_IT8XXX2_BITRATE_1MHZ:
+		hoctl2r_to_write |= PECI_IT8XXX2_BITRATE_1MHZ_BITS;
+		break;
+
+	case PECI_IT8XXX2_BITRATE_1P6MHZ:
+		hoctl2r_to_write |= PECI_IT8XXX2_BITRATE_1P6MHZ_BITS;
+		break;
+
+	default:
+		printk("[PECI] Error: Specified Bitrate Not Supported\r\n");
```
Comment (keith-zephyr) [inline drivers/peci/peci_ite_it8xxx2.c:209]: Use logging here instead.  Also return -ENOTSUP.

File: drivers/peci/peci_ite_it8xxx2.c [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+static void peci_it8xxx2_rst_module(const struct device *dev)
+{
+	const struct peci_it8xxx2_config *config = dev->config;
+	struct peci_it8xxx2_regs *const peci_regs =
+		(struct peci_it8xxx2_regs *)config->base_addr;
+
+	peci_regs->HOSTAR = HOSTAR_RST_ANYBIT;
+
+	printk("[PECI] Module Reset for Status Error.\r\n");
```
Comment (keith-zephyr) [inline drivers/peci/peci_ite_it8xxx2.c:249]: Use logging.

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts (hunk 1 above)
Comment (itebj) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:18]: Dear Marti, 

Thanks for response.
The alias is added for the sample code. According to the line 190 in the main.c which can be found in the directory "samples/drivers/peci/src":

        peci_dev = DEVICE_DT_GET(DT_ALIAS(peci_0));

In my personal test code, I did use the DT_NODELABEL instead of ALIAS.
Besides, the reason why there's an i2c-0, I believe that it's from another driver module which could probably encounter the same scenario.

Thanks again.

Best regards,
BJ

File: drivers/peci/peci_ite_it8xxx2.c (hunk 2 above)
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:364]: Dear Marti,

Thanks for reminding me of this. Hmm, I'd like to change it but when I read the specification of PECI3.0 and the Intel platform datasheets, it seems that there's always 1-pin used in the PECI interface. Does it mean that there's probability that the PECI be extended? 

Thanks for the patience.

Best regards,
BJ

File: drivers/peci/peci_ite_it8xxx2.c [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+/* HOSTAR (F02C00h) */
+#define HOBY			BIT(0)
+#define FINISH			BIT(1)
+#define RD_FCS_ERR		BIT(2)
+#define WR_FCS_ERR		BIT(3)
+#define EXTERR			BIT(5)
+#define BUS_ER			BIT(6)
+#define TEMPERR			BIT(7)
+#define HOSTAR_RST_ANYBIT \
+		(TEMPERR+BUS_ER+EXTERR+WR_FCS_ERR+RD_FCS_ERR+FINISH)
```
[the remark this answers is not in the document]
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:53]: Dear Keith,

Thanks for correcting me that I really did a rookie mistake here.
Sorry about that and quite appreciative of your guidance.
I'll correct it.

Best regards,
BJ

File: drivers/peci/peci_ite_it8xxx2.c (hunk 3 above)
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:209]: Got it, sir.
I'll change it to logging from now on.

Truly grateful.

File: drivers/peci/peci_ite_it8xxx2.c (hunk 4 above)
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:249]: Understood.

File: drivers/peci/peci_ite_it8xxx2.c (hunk 2 above)
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:364]: Dear Marti,

Thanks for your guidance, and I'd like to consult you with if the following statement is appropriate to describe the current situation of this driver module.

`
/*
 * This driver is single-instance. If the devicetree contains multiple
 * instances, this will fail and the driver needs to be revisited.
 */
BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) <= 1,
	     "Unsupported PECI Instance");
`

Thanks again for your helps.

File: drivers/peci/peci_ite_it8xxx2.c [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+	.base_addr = DT_INST_REG_ADDR(0),
+	.irq_no = DT_INST_IRQN(0),
+	.alts_list = peci_alt_inst,
+};
+
+static struct peci_it8xxx2_data peci_it8xxx2_data0;
+
+/* ITE IT8XXX2 PECI Functions */
+
+static void peci_it8xxx2_set_dvie(const struct device *dev, bool en)
```
[the remark this answers is not in the document]
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c]: Dear Keith, 

Thanks for guidance.
The function `peci_it8xxx2_set_dvie` has been removed because it's true that it should be set once in the initialization. 
Secondly, when it comes to passing a pointer to peci_regs instead of passing `dev`, I've checked some private functions and made some modifications. Please advise if there's any further suggestion.

Thanks again.
BJ

File: drivers/peci/peci_ite_it8xxx2.c (hunk 2 above)
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:349]: Hi Marti,

Thanks for sharing your opinions.
I've added the following statements in the latest code:

```
/*
 * This driver is single-instance. If the devicetree contains multiple
 * instances, this will fail and the driver needs to be revisited.
 */
BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) <= 1,
	     "Unsupported PECI Instance");
```

Meanwhile, so far the latest image is also tested with the `samples/drivers/peci`.
Thanks for guidance.
BJ

File: drivers/peci/peci_ite_it8xxx2.c [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+LOG_MODULE_REGISTER(peci_ite_it8xxx2, CONFIG_PECI_LOG_LEVEL);
+
+/*
+ * This driver is single-instance. If the devicetree contains multiple
+ * instances, this will fail and the driver needs to be revisited.
+ */
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) <= 1,
+	     "Unsupported PECI Instance");
+
```
Comment (keith-zephyr) [inline drivers/peci/peci_ite_it8xxx2.c:29]: This driver currently requires interrupts, so you should check that `CONFIG_PECI_INTERRUPT_DRIVEN` is enabled.

File: soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series [hunk 8]
```diff
@@ -76,6 +76,10 @@ config TACH_IT8XXX2
 	default y
 	depends on SENSOR
 
+config PECI_ITE_IT8XXX2
+	default y
+	depends on PECI
```
Comment (keith-zephyr) [inline soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series:81]: Add `select PECI_INTERRUPT_DRIVEN` because the driver requires it.

File: drivers/peci/peci_ite_it8xxx2.c [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+{
+	const struct peci_it8xxx2_config *config = dev->config;
+	struct peci_it8xxx2_regs *const peci_regs =
+		(struct peci_it8xxx2_regs *)config->base_addr;
+
+	peci_regs->HOSTAR = HOSTAR_RST_ANYBIT;
+
+	LOG_ERR("[PECI] Module Reset for Status Error.\r\n");
+	/* Reset IT8XXX2 PECI Module Thoroughly */
+	IT83XX_GCTRL_RSTC4 |= RPECI;
```
Comment (keith-zephyr) [inline drivers/peci/peci_ite_it8xxx2.c:234]: The Chromium EC driver included a [delay](https://chromium.googlesource.com/chromiumos/platform/ec/+/01aa26ee/chip/it83xx/peci.c#66) after this reset.  Is that still required?

File: drivers/peci/peci_ite_it8xxx2.c [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+	return 0;
+}
+
+static void peci_it8xxx2_rst_module(const struct device *dev)
+{
+	const struct peci_it8xxx2_config *config = dev->config;
+	struct peci_it8xxx2_regs *const peci_regs =
+		(struct peci_it8xxx2_regs *)config->base_addr;
+
+	peci_regs->HOSTAR = HOSTAR_RST_ANYBIT;
```
Comment (keith-zephyr) [inline drivers/peci/peci_ite_it8xxx2.c:230]: Replace with `peci_it8xxx2_rst_status(peci_regs);`

Also, should the interrupt status bits be cleared after the global reset at line 234 instead?

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: drivers/peci/peci_ite_it8xxx2.c (hunk 7 above)
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:29]: Truly appreciative of your guidance, I'll add it ASAP. :-D

File: soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series (hunk 8 above)
Comment (itebj) [reply soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series:81]: Understood.

File: drivers/peci/peci_ite_it8xxx2.c (hunk 9 above)
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:234]: So far I run the test code with the IT81302, and it seems working without putting a delay. I'll make more experiments on this part.

File: drivers/peci/peci_ite_it8xxx2.c (hunk 10 above)
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:230]: When it comes to the global reset, I think that the BIOS would reinitialize the PECI module in the POST. However, if the transaction failed to work, it definitely would lead the code to the function `peci_it8xxx2_rst_module ` which resets the PECI module and I believe that the status register has been cleared before module reset. Besides, we also believe that all the PECI regs would be cleared thoroughly when the module is reset by writing the corresponding reset bit to the register RSTC4. 

```
	peci_regs->HOSTAR = HOSTAR_RST_ANYBIT;

	LOG_ERR("[PECI] Module Reset for Status Error.\r\n");
	/* Reset IT8XXX2 PECI Module Thoroughly */
	IT83XX_GCTRL_RSTC4 |= RPECI;
```

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts (hunk 1 above)
Comment (itebj) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:18]: Thanks for the URL, sir. And yes, it's the sample code which I mentioned and built with. So is it required to remove the alias? Does it mean that we need to replace all the DT_ALIAS with DT_NODELABEL in the future?

File: drivers/peci/peci_ite_it8xxx2.c (hunk 7 above)
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:32]: Added the following statements:

```
BUILD_ASSERT(IS_ENABLED(CONFIG_PECI_INTERRUPT_DRIVEN),
	     "Please enable the option CONFIG_PECI_INTERRUPT_DRIVEN");
```

File: soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series (hunk 8 above)
Comment (itebj) [reply soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series:81]: Added as follows:

```
config PECI_ITE_IT8XXX2
	default y
	depends on PECI
	select PECI_INTERRUPT_DRIVEN
```

**Review by jackrosenthal — APPROVED** [review]

File: drivers/peci/peci_ite_it8xxx2.c (hunk 10 above)
Comment (keith-zephyr) [reply drivers/peci/peci_ite_it8xxx2.c:232]: If setting RPECI clears the pending interrupts, you should delete line 232.

If an explicit clear of the interrupts is needed, pplease change line 232 to call `peci_it8xxx2_rst_status(peci_regs);`

Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:232]: Dear Keith, 

So appreciative of your guidance, yes I should delete line 232.
Thanks again.

File: drivers/peci/peci_ite_it8xxx2.c (hunk 9 above)
Comment (itebj) [reply drivers/peci/peci_ite_it8xxx2.c:234]: So far we've run it for days, it looks good without the delay after resets.

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts (hunk 1 above)
Comment (mbolivar-nordic) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:18]: Thank you -- this makes sense. I appreciate the explanation.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by keith-zephyr — APPROVED** [review]

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts (hunk 1 above)
Comment (itebj) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:18]: Thanks again for your guidance and patience. Truly grateful. :-D
